addon/-private/system/store.js

```diff
@@ -2417,20 +2419,63 @@ Store = Service.extend({
     heimdall.increment(_setupRelationships);
     let setupToken = heimdall.start('store._setupRelationships');
     let pushed = this._pushedInternalModels;
+    let store = this;
 
     // Cache the inverse maps for each modelClass that we visit during this
     // payload push.  In the common case where we are pushing many more
     // instances than types we want to minimize the cost of looking up the
     // inverse map and the overhead of Ember.get adds up.
     let modelNameToInverseMap = Object.create(null);
+    let modelNameToRelationshipsMap = Object.create(null);
 
-    for (let i = 0, l = pushed.length; i < l; i += 2) {
-      // This will convert relationships specified as IDs into DS.Model instances
-      // (possibly unloaded) and also create the data structures used to track
-      // relationships.
+    for (let i = 0; i < pushed.length; i += 2) {
       let internalModel = pushed[i];
-      let data = pushed[i + 1];
-      setupRelationships(this, internalModel, data, modelNameToInverseMap);
+      let data = pushed [i + 1];
+      let modelName = internalModel.modelName;
+      let relationshipsByName = modelNameToRelationshipsMap[modelName];
+
+      if (relationshipsByName === undefined) {
+        relationshipsByName = modelNameToRelationshipsMap[modelName] = get(internalModel.modelClass, 'relationshipsByName');
+      }
+
+      let relationshipNames = relationshipsByName._keys.list;
+      let relationships = internalModel._relationships;
+
+      for (let keyIndex = 0; keyIndex < relationshipNames.length; keyIndex++) {
+        let relationshipName = relationshipNames[keyIndex];
+        let relationshipData = data.relationships[relationshipName];
+
+        if (relationshipData !== undefined) {
+          let relationshipRequiresNotification = relationships.has(relationshipName) ||
+            isInverseRelationshipInitialized(store, internalModel, data, relationshipName, modelNameToInverseMap);
+
+          if (relationshipRequiresNotification) {
+            relationships.get(relationshipName).push(relationshipData);
+          }
+
+          // in debug, assert payload validity eagerly
+          runInDebug(() => {
```

Comment:
If you'd strip the `runInDebug` before running Babel on it, then Babel might be able to avoid the closures.